File: Engine/src/main/java/org/etieskrill/engine/graphics/texture/animation/TextureAnimationYamlParser.java
package org.etieskrill.engine.graphics.texture.animation;

import lombok.NoArgsConstructor;
import org.etieskrill.engine.config.ResourcePaths;
import org.joml.Vector2i;
import org.joml.primitives.Rectanglei;
import org.yaml.snakeyaml.Yaml;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReference;

import static java.lang.Math.max;
import static org.etieskrill.engine.util.ResourceReader.getClasspathResourceAsStream;

public class TextureAnimationYamlParser {

    private static final float MILLIS_TO_SECONDS = 0.001f;

    static TextureAnimationMetadata loadAnimationMetadata(String metaFile) {
        Yaml yaml = new Yaml();
        YamlTextureAnimationMetadata yamlMetaData = yaml.loadAs(
                getClasspathResourceAsStream(ResourcePaths.TEXTURE_PATH + metaFile),
                YamlTextureAnimationMetadata.class
        );
        if (yamlMetaData.resources == null || yamlMetaData.resources.get("texture") == null) {
            throw new RuntimeException("Animation resource cannot be null");
        }

        List<TextureAnimationFrame> frames = new ArrayList<>();
        YamlTextureAnimationFrame defaultFrame = yamlMetaData.frames.get("default");
        AtomicInteger maxX = new AtomicInteger(), maxY = new AtomicInteger();
        AtomicReference<Float> currentTimeMillis = new AtomicReference<>(0f);
        yamlMetaData.frames.forEach((frameName, frame) -> {
            if (frameName.equals("default")) return;
            var parsedFrame = parseFrame(frame, defaultFrame, currentTimeMillis);
            maxX.set(max(maxX.get(), parsedFrame.getAtlasArea().lengthX()));
            maxY.set(max(maxY.get(), parsedFrame.getAtlasArea().lengthY()));
            frames.add(parsedFrame);
        });

        return new TextureAnimationMetadata(
                yamlMetaData.resources.get("texture").file,
                new Vector2i(maxX.get(), maxY.get()),
                null, //TODO
                frames,
                currentTimeMillis.get() * MILLIS_TO_SECONDS
        );
    }

    private static TextureAnimationFrame parseFrame(YamlTextureAnimationFrame frame, YamlTextureAnimationFrame defaultFrame, AtomicReference<Float> currentTimeMillis) {
        if (frame == null) {
            return parseDefaultFrame(defaultFrame, currentTimeMillis);
        } else {
            var newFrame = new TextureAnimationFrame(
                    new Rectanglei(getElse(frame.x, defaultFrame.x), getElse(frame.y, defaultFrame.y),
                            getElse(frame.x, defaultFrame.x) + getElse(frame.w, defaultFrame.w),
                            getElse(frame.y, defaultFrame.y) + getElse(frame.h, defaultFrame.h)),
                    currentTimeMillis.get() * MILLIS_TO_SECONDS);
            currentTimeMillis.updateAndGet(currentTime -> currentTime + getElse(frame.duration, defaultFrame.duration));
            return newFrame;
        }
    }

    private static TextureAnimationFrame parseDefaultFrame(YamlTextureAnimationFrame yamlFrame, AtomicReference<Float> currentTimeMillis) {
        var frame = new TextureAnimationFrame(
                new Rectanglei(yamlFrame.x, yamlFrame.y, yamlFrame.x + yamlFrame.w, yamlFrame.y + yamlFrame.h),
                currentTimeMillis.get() * MILLIS_TO_SECONDS);
        currentTimeMillis.updateAndGet(currentTime -> currentTime + yamlFrame.duration);
        return frame;
    }

    private static <T> T getElse(T object, T defaultObject) {
        return object == null ? defaultObject : object;
    }

}

@NoArgsConstructor
class YamlTextureAnimationMetadata {
    public String $schema;
    public Map<String, YamlTextureAnimationResource> resources;
    public Map<String, YamlTextureAnimationFrame> frames;
}

@NoArgsConstructor
class YamlTextureAnimationResource {
    public String file;
}

@NoArgsConstructor
class YamlTextureAnimationFrame {
    public Integer x, y, w, h;
    public Float duration;
}

File: Engine/src/commonMain/kotlin/org/etieskrill/engine/graphics/texture/animation/TextureAnimationYamlParser.kt
package org.etieskrill.engine.graphics.texture.animation

import org.etieskrill.engine.graphics.texture.TextureFormat
import org.etieskrill.engine.util.readResource
import org.joml.Vector2i
import org.joml.primitives.Rectanglei
import org.yaml.snakeyaml.Yaml

private const val MILLIS_TO_SECONDS = 0.001f

data class YamlTextureAnimationMetadata(
    val `$schema`: String,
    val resources: Map<String, YamlTextureAnimationResource>,
    val frames: Map<String, YamlTextureAnimationFrame>
)

data class YamlTextureAnimationResource(val file: String)

data class YamlTextureAnimationFrame(
    val x: Int?,
    val y: Int?,
    val w: Int?,
    val h: Int?,
    val duration: Float?
)

internal fun loadAnimationMetadata(metaFile: String, format: TextureFormat): TextureAnimationMetadata {
    val yaml = Yaml()
    val yamlMetaData = yaml.loadAs(readResource(metaFile).inputStream(), YamlTextureAnimationMetadata::class.java)

    check(yamlMetaData.resources["texture"] != null) { "Animation resource cannot be null" }

    val defaultFrame = yamlMetaData.frames["default"] ?: error("Animated texture must have a default frame")
    var maxX = 0
    var maxY = 0
    var currentTimeMillis = 0f

    val frames = yamlMetaData.frames
        .filterNot { it.key == "default" }
        .values.map { frame ->
            val parsedFrame = parseFrame(frame, defaultFrame, currentTimeMillis)
            maxX = maxOf(maxX, parsedFrame.atlasArea.lengthX())
            maxY = maxOf(maxY, parsedFrame.atlasArea.lengthY())
            currentTimeMillis += frame.duration ?: defaultFrame.duration!!
            parsedFrame
        }

    return TextureAnimationMetadata(
        yamlMetaData.resources["texture"]!!.file,
        Vector2i(maxX, maxY),
        format,
        frames,
        currentTimeMillis * MILLIS_TO_SECONDS
    )
}

private fun parseFrame(
    frame: YamlTextureAnimationFrame?,
    defaultFrame: YamlTextureAnimationFrame,
    currentTimeMillis: Float
): TextureAnimationFrame {
    frame ?: return parseDefaultFrame(defaultFrame, currentTimeMillis)

    val x = frame.x ?: defaultFrame.x!!
    val y = frame.y ?: defaultFrame.y!!
    val newFrame = TextureAnimationFrame(
        Rectanglei(x, y, x + (frame.w ?: defaultFrame.w!!), y + (frame.h ?: defaultFrame.h!!)),
        currentTimeMillis * MILLIS_TO_SECONDS
    )

    return newFrame
}

private fun parseDefaultFrame(yamlFrame: YamlTextureAnimationFrame, currentTimeMillis: Float) =
    TextureAnimationFrame(
        Rectanglei(yamlFrame.x!!, yamlFrame.y!!, yamlFrame.x + yamlFrame.w!!, yamlFrame.y + yamlFrame.h!!),
        currentTimeMillis * MILLIS_TO_SECONDS
    )
